### PR TITLE
Fix SR-15785 by adding a cached reference to the last linked list node, avoiding n^2 append scaling

### PR DESCRIPTION
**Pull request by Catfish-Man**
Speeds up a trivial benchmark with 50,000 appends from 10+ seconds in a release build, to 0.2 seconds in a debug build, on my machine.

Fixes rdar://88219931

### CODE REVIEW

**Review by ktoso — APPROVED** [review]
Sweet, LGTM 👍

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by ogres** [comment]
it would be nice, if we could also add a benchmark to ensure that we do not regress in future

**Comment by Catfish-Man** [comment]
Yes, I'm working on that currently. Using Concurrency stuff in the benchmarks is slightly non-obvious for now.

**Comment by Catfish-Man** [comment]
Turns out "slightly non-obvious" is actually more like extremely non-obvious, adding a regression test will have to wait a bit, sorry.

File: include/swift/ABI/TaskStatus.h [hunk 1]
```diff
@@ -164,14 +164,19 @@ class ChildTaskStatusRecord : public TaskStatusRecord {
...
+      : TaskStatusRecord(TaskStatusRecordKind::TaskGroup),
+        FirstChild(nullptr),
+        LastChild(nullptr) {
   }
 
   TaskGroupTaskStatusRecord(AsyncTask *child)
-      : TaskStatusRecord(TaskStatusRecordKind::TaskGroup), FirstChild(child) {}
+      : TaskStatusRecord(TaskStatusRecordKind::TaskGroup),
+        FirstChild(child),
+        LastChild(child) {}
```
Comment (rjmccall) [inline include/swift/ABI/TaskStatus.h:179]: Can you assert `!LastChild || !LastChild->childFragment()->getNextChild()` here?

Comment (Catfish-Man) [reply include/swift/ABI/TaskStatus.h:179]: Sure

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Pull request merged**